engines/ic3ia.cpp

```diff
@@ -0,0 +1,369 @@
+
+  // check if last frame intersects with bad
+  assert_frame_labels(reached_k_ + 1);
+  solver_->assert_formula(bad_);
+  Result r = solver_->check_sat();
+
+  if (r.is_sat()) {
+    add_proof_goal(get_conjunction_from_model(), reached_k_ + 1, NULL);
+  }
+
+  pop_solver_context();
+
+  return r.is_sat();
+}
+
+bool IC3IA::get_predecessor(size_t i,
+                            const Conjunction & c,
+                            Conjunction & out_pred)
+{
+  assert(i > 0);
+  assert(i < frames_.size());
+
+  assert(solver_context_ == 0);
+  push_solver_context();
+
+  // F[i-1]
+  assert_frame_labels(i - 1);
+  // -c
+  solver_->assert_formula(solver_->make_term(Not, c.term_));
+  // Trans
+  assert_trans_label();
+  // c'
+  solver_->assert_formula(next(c.term_));
+
+  Result r = solver_->check_sat();
+  if (r.is_sat()) {
+    // TODO: add generalize option
+    // TODO: refactor mbic3 to use get_conjunction_from_model and we won't need
+    // to override it
+    out_pred = get_conjunction_from_model();
+    pop_solver_context();
+  } else {
+    pop_solver_context();
+    assert(solver_context_ == 0);
+    TermVec assump, red_assump, rem_assump;
+    for (auto a : c.conjuncts_) {
+      assump.push_back(abs_ts_.next(a));
+    }
+
+    Term formula = make_and(TermVec{
+        get_frame(i - 1), solver_->make_term(Not, c.term_), trans_label_ });
+
+    // filter using unsatcore
+    reduce_assump_unsatcore(formula, assump, red_assump, &rem_assump);
+    // get current version of red_assump
+    TermVec cur_red_assump, cur_rem_assump;
+    for (auto a : red_assump) {
+      cur_red_assump.push_back(abs_ts_.curr(a));
+    }
+    for (auto a : rem_assump) {
+      cur_rem_assump.push_back(abs_ts_.curr(a));
+    }
+
+    fix_if_intersects_initial(cur_red_assump, cur_rem_assump);
+    assert(cur_red_assump.size() > 0);
+    out_pred = Conjunction(solver_, cur_red_assump);
+  }
+
+  assert(!r.is_unknown());
+  return r.is_sat();
+}
+
+Conjunction IC3IA::get_conjunction_from_model()
+{
+  TermVec conjuncts;
+  conjuncts.reserve(pred_statevars_.size());
+  Term val;
+  for (auto p : pred_statevars_) {
+    if ((val = solver_->get_value(p)) == true_) {
+      conjuncts.push_back(p);
+    } else {
+      assert(val == false_);
+      conjuncts.push_back(solver_->make_term(Not, p));
+    }
+  }
+  return Conjunction(solver_, conjuncts);
+}
+
+void IC3IA::set_labels()
+{
+  assert(solver_context_ == 0);  // expecting to be at base context level
+  // set semantics of labels
+  Sort boolsort = solver_->make_sort(BOOL);
+  if (!init_label_) {
+    // frame 0 label is identical to init label
+    init_label_ = frame_labels_[0];
+  }
+  if (!trans_label_) {
+    trans_label_ = solver_->make_symbol("__trans_label", boolsort);
+    // Note: Using iff instead of implies so that we can negate it
+    // for example in relational preimage generalization
+    solver_->assert_formula(
+        solver_->make_term(Iff, trans_label_, abs_ts_.trans()));
```

Comment:
we changed it to `implies` in the master.

Reply:
shouldn't we also change here?

Reply:
Yeah good catch -- I missed it. Updated with 54e716aabee3cff707cfdc5a22c601530234616a.